### DISCUSSION TITLE
Install an older version of Homebrew on macOS 10.13 High Sierra [solved]

### DISCUSSION THREAD

**Discussion by h-archer** [Homebrew/discussions#1487] [Everyday usage]
Hi,

I have to deliver my finished website today to my client, but on this day, on my old laptop, that is only able to go to macOS 10.13 High Sierra, I made the MASSIVE mistake of running "brew upgrade"!!! 

It ran for 2 hrs. Then, broke everything on my local dev environment Completely! Php, Valet, nginx, its all not running now... Spent the whole days searching but cannot find a way to revert homebrew back.

Now I am STUCK!! With a client that wants to have his website!

So I uninstalled homebrew. Now I need to install Homebrew version 2.2.12, which seems to be the last compatible version with macOS 10.13 High Sierra. 

How do I install this version of Homebrew Please ([version 2.2.12](https://github.com/Homebrew/brew/releases/tag/2.2.12))? Can anyone help please?

**Comment by carlocab** [comment]
`brew` code is all in `git`, so going back to an earlier version just involves checking out the right tag. However, the bigger challenge is going back in time to find the bottles that worked on High Sierra. Those all lived on Bintray, but Bintray is now gone.

**Reply by h-archer** [reply]
Thanks @carlocab. I tried git checking to an older release using the advise from this [issue](https://stackoverflow.com/questions/62049367/how-can-i-downgrade-homebrew-to-an-older-version)

But that did not work so I resorted to uninstalling Homebrew so I can reinstall from fresh.

The question is, from [this page](https://github.com/Homebrew/brew/releases/tag/2.2.12) I can get the source for 2.2.12. Is there a way I can install this manually? The issue is that I need homebrew because I am relying on packaging that need it. Macports is not an option also. There must be a way I can install homebrew 2.2.12.

**Reply by carlocab** [reply]
Installing `brew` 2.2.12 is the same as installing the latest version of Homebrew. You install `brew` then checkout the right tag from the Homebrew repository. That said, I don't think checking out an older version of `brew` will fix your problems.

**Reply by h-archer** [reply]
Thanks again mate. Yes I think you are right. Because after the upgrade, I did a checkout to version 2.2.12 and still had issues. Since you are an expert and I need an urgent solution, what do you recommend so that I can get my local website working on my laptop again?

**Reply by carlocab** [reply]
Unfortunately I don't know of any solution that doesn't involve upgrading to Mojave, Catalina, or Big Sur.

**Reply by h-archer** [reply]
Which means I need to look at another solution other than brew. Thank for your help! I don't know what to do now :(

**Reply by gromgit** [reply]
@hga77 You could also restore from a Time Machine backup. (If you don't have one, you've just learned the hard way about the importance of proper backups.)

**Reply by carlocab** [reply]
While you're working that out, I'd try identifying a minimal set of packages you absolutely need to get running and try to `brew install` those using the latest version of `brew`. You'll need to build them from source, which will take some time. `brew doctor` should show you things you can do to try to improve the success of source builds.

If you can get `brew doctor` to a state that says `Your system is ready to brew.`, then you should be able to install everything that works on High Sierra. (Sadly not all formulae install on High Sierra.)

**Reply by h-archer** [reply]
@gromgit You know I was thinking of doing a manual time machine backup last week but did not. You're right I have just learned the hard way!

@carlocab Thanks for the advise. Just to get a better understanding. Are these the following steps to carry out?

1. Install brew using the instruction on the brew homepage. This will install latest brew version.

2. I need php7.3, mariadb, valet (includes nginx). That's what I need to get the website running locally on the laptop. Now these packages are already on my laptop and configured. For example, php is still at the same destination and has my configurations. So I don't want to re-install php (or valet, nginx). How do I simply enable them on the new brew? Because if I install these packages a) they will be latest versions and b) they won't have my configs.

**Reply by h-archer** [reply]
Just installed brew and its on 3.1.8. And here's what brew doctor returns:

```console
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/wkhtmltox/image.h
  /usr/local/include/wkhtmltox/pdf.h

Warning: Unbrewed '.la' files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.la' files:
  /usr/local/lib/libhtsjava.la
  /usr/local/lib/libhttrack.la

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libhtsjava.a
  /usr/local/lib/libhttrack.a

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run brew link on these:
  nginx
  dnsmasq

Warning: You are using macOS 10.13.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.


Warning: Some installed formulae are missing dependencies.
You should brew install the missing dependencies:
  brew install apr apr-util argon2 aspell autoconf brotli c-ares curl freetds freetype gdbm gettext glib gmp icu4c jemalloc jpeg krb5 libev libffi libidn2 libmetalink libpng libpq libsodium libssh2 libtiff libtool libunistring libzip m4 mpdecimal nghttp2 openldap openssl@1.1 pcre python@3.9 readline rtmpdump sqlite tidy-html5 unixodbc webp xz zstd

Run brew missing for more details.
```


How can I resolve these warnings?

**Reply by SMillerDev** [reply]
> Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

If your install is otherwise fine you can just ignore the warnings. `brew install php7@.3 mariadb valet` should install what you want.

**Reply by h-archer** [reply]
Thanks @SMillerDev 
My only concern is that if I re-install mariadb, will this deleted my current databases?
And if I re-install php7.3, will I loose my php.ini configs?

Can't I just re-enable php/nginx/valet/mariadb? because they are installed but I just need to activate them with the new brew.

**Reply by SMillerDev** [reply]
`brew install` won't delete data, `brew reinstall` might.

> Can't I just re-enable php/nginx/valet/mariadb? because they are installed but I just need to activate them with the new brew.

brew doesn't know an "enable" concept. Something is either installed or not. `brew ls` will tell you what's installed.

**Reply by pacman-admin** [reply]
same here!!!
I have a 21.5 inch iMac 2011 and running brew upgrade broke EVERYTHING!!!
Openldap simply refuses to upgrade and curl has weird errors.

**Comment by h-archer** [comment]
So homebrew is now latest version 3.1.8 and brew doctor has no warnings (other than the message that my macos is old).

I tried to install php but I get the following error (image below) during the install. It also asks for a keychain password for github in a popup. Not sure what that is about. Same error when trying to install php7.3, 7.4 and 8. Also when trying to install mariadb.

![Screen-Shot-2021-05-19-at-02-56-30-png-1135×623-](https://user-images.githubusercontent.com/396065/118738130-fdcd9280-b84e-11eb-92ba-cad484cb7176.png)

What should I try next? Why doesn't php even install??

**Reply by gromgit** [reply]
I can confirm that the current `cmake` formula does _not_ install from source even on my Intel Big Sur box, but work commitments prevent me from digging any further for now.

**Reply by h-archer** [reply]
Thanks @gromgit. Can someone have a dig at this? Otherwise where should I report this issue?

**Reply by SMillerDev** [reply]
https://github.com/Homebrew/homebrew-core/pull/77596 should resolve the build issue.

**Reply by h-archer** [reply]
> [Homebrew/homebrew-core#77596](https://github.com/Homebrew/homebrew-core/pull/77596) should resolve the build issue.

Thanks again @SMillerDev
What do I need to do to use this patch?

**Reply by SMillerDev** [reply]
Wait until it's merged is by far the best option

**Reply by h-archer** [reply]
@SMillerDev Just to confirm, what is the estimated time to merging?

**Reply by SMillerDev** [reply]
Whenever someone has time to merge it and CI is green.

**Reply by h-archer** [reply]
@SMillerDev btw thanks for creating the patch. I hope this enables me to install php/valet etc...

**Reply by h-archer** [reply]
@SMillerDev, One test just failed and I urgently need this patch on my machine. Is there a way to manually apply this patch and not have to wait?

**Reply by carlocab** [reply]
There is, but it requires some proficiency with GitHub. The `gh` command-line tool would be helpful here. However, I've triggered a merge for https://github.com/Homebrew/homebrew-core/pull/77596, and it should be available ~in a few minutes~ now. You'll need to `brew update`.

**Reply by h-archer** [reply]
Thanks @carlocab, much appreciated!

php@7.3 has installed :). Now I have the same issue with mariadb. Please see my post below.

**Comment by h-archer** [comment]
I have just managed to install php@7.3 successfully thanks to the patch from @SMillerDev and the support from the other guys. Thank you.

Now I am having the same issue with mariadb. I did a `brew update` followed by `brew install mariadb`. It failed at the last step while installing mariadb (please see image). Does this also need a patch for mariadb?

![Screen-Shot-2021-05-19-at-18-21-12-png-1175×831-](https://user-images.githubusercontent.com/396065/118851336-4da56b80-b8da-11eb-894e-cd4b050b952d.png)

**Reply by carlocab** [reply]
I've triggered a [build](https://github.com/carlocab/homebrew-rpath-test/pull/32/checks?check_run_id=2622420345) to check.

By the way, please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse (e.g. for Googling error messages).

**Reply by h-archer** [reply]
Thank you @carlocab for sharing that link. Makes sense not to use images. Noted.

I hope you find something from the build you just ran!

**Reply by carlocab** [reply]
Yup, fails with exactly the same error: https://github.com/carlocab/homebrew-rpath-test/pull/32/checks?check_run_id=2623220118#step:8:766

Probably broken by the new CMake. Needs either a patch or a version bump.

**Reply by h-archer** [reply]
Thanks again @carlocab. You guys are very helpful. 

I guess it's a waiting game now for someone to fix this?

**Reply by carlocab** [reply]
Yes. I found a patch that could help: https://github.com/mariadb-corporation/mariadb-connector-c/commit/242cab8cbcd91af882233730a83627d3b12ced83

Not sure if that's all that's needed though.

**Reply by h-archer** [reply]
@carlocab Sorry for my noob question but how do I apply it? Does someone need to merge it like the other patch?

**Reply by carlocab** [reply]
You can try applying this change to your local copy of `mariadb`:
```diff
diff --git a/Formula/mariadb.rb b/Formula/mariadb.rb
index c5ec1ff27e..3e20b4690b 100644
--- a/Formula/mariadb.rb
+++ b/Formula/mariadb.rb
@@ -36,6 +36,14 @@ class Mariadb < Formula
   conflicts_with "mytop", because: "both install `mytop` binaries"
   conflicts_with "mariadb-connector-c", because: "both install `mariadb_config`"
 
+  # Fix syntax error with CMake 3.20
+  # https://github.com/mariadb-corporation/mariadb-connector-c/commit/242cab8cbcd91af882233730a83627d3b12ced83
+  patch do
+    url "https://github.com/mariadb-corporation/mariadb-connector-c/commit/242cab8cbcd91af882233730a83627d3b12ced83.patch?full_index=1"
+    sha256 "760fd19cd8d4d756a0799ed9110cfd2898237e43835fefe3668079c5b87fc36d"
+    directory "libmariadb"
+  end
+
   def install
     # Set basedir and ldata so that mysql_install_db can find the server
     # without needing an explicit path to be set. This can still
```
You can edit your local copy with `brew edit mariadb`. (I hope you're well-versed with Git, or making these local changes can turn into a headache.)

**Reply by h-archer** [reply]
I use Git but not sure if I have what it takes to do this. I will give it a go. Thanks again :)

**Reply by carlocab** [reply]
If you ever want to start from scratch: `brew update-reset`

Note, however, this will likely destroy any changes you've made to your local repo (that you haven't committed to a branch that isn't master).

**Reply by h-archer** [reply]
Thank you so much @carlocab. You'll be glad to know that I made the change to mariadb, as you instructed, and it has installed successfully! And the website that I was working on is RUNNING locally again!!!! You guys have saved me and now I can carry on finishing my work!

Thanks to @carlocab, @SMillerDev and @gromgit. I don't think I'll be doing a `brew upgrade` any time soon that's for sure!

**Reply by carlocab** [reply]
Glad you're up and running again.

You might want to try `brew pin`:

```
❯ brew help pin
Usage: brew pin installed_formula [...]

Pin the specified formula, preventing them from being upgraded when issuing
the brew upgrade formula command. See also unpin.

  -d, --debug                      Display any debugging information.
  -q, --quiet                      Make some output more quiet.
  -v, --verbose                    Make some output more verbose.
  -h, --help                       Show this message.
```

And please consider upgrading to a laptop running an OS that Apple still supports :)

**Reply by h-archer** [reply]
Thanks for that tip about pinning a formula! I will look into that. 

> And please consider upgrading to a laptop running an OS that Apple still supports :)

This experience has made me realise that it might be time to upgrade. Definitely.

**Reply by gromgit** [reply]
> This experience has made me realise that it might be time to upgrade. Definitely.

And set up some sort of backup. Helping others recover from lost data is something that's not part of $DAYJOB, but I end up doing it far too often regardless.

**Reply by h-archer** [reply]
Yes @gromgit. I am very lucky this time. I will be planning to upgrade and have a backup setup. 

**Let this issue be a reminder for others to have a backup setup in place.**

**Comment by iguvenilir** [comment]
So What is actually the procedure to install brew on mac high sierra ?
Can somebody put instructions 1, 2, 3. please.

**Reply by gromgit** [reply]
1. Follow the instructions on the [Homebrew home page](https://brew.sh/) under **Install Homebrew**.
2. If the installer doesn't work for some reason, try the **Untar anywhere** method documented [here](https://docs.brew.sh/Installation#untar-anywhere-unsupported).

Once either of the above completes without error, you now face the hardest part of all: building every formula you need _and all their dependencies_ from source. Some formulae may no longer build under High Sierra for various reasons; you'll have to figure out what to do then.

**Reply by konnano** [reply]
In order to use home brew in 10.13, it is necessary to devise installation for each formula
